(#384) added pull progress report via event OfflineDbContext.SynchronizationProgress

StefanCuypers opened the pull request:
OfflineDbContext.SynchronizationProgress event is added to provide progress report during pull operations. Both source fetching and local committing progress are reported.

adrianhall commented:
This looks good - I'll take the compile checks on later this week.

Do you want to add a PullOperationStart and PullOperationEnd event as well?  I'm ambivalent to the PullOperationStart, but I think it would be useful for the async process to understand when they won't be getting more events.

adrianhall reviewed (changes requested):
General comment - should we also report progress on Push?

The AI caught what is likely a logical bug in the test as well, which should be fixed or explained before merging.

StefanCuypers commented:
I've added pull start and end events and included Exception reporting in the end event, which could be used to have the info in request #383 .
I also added reporting on push operations. Used the same event and adjusted it since the nature of push and pull is a bit different.  On pull we report per entity type, where on push we do not since pushes work historicaly rather than by entity type.
And I fixed the test bug and added tests for Exception reporting on pull and for the push event reporting.

StefanCuypers commented:
@adrianhall I don't know why the Create_CanRoundtrip_Types test suddenly fails and worked before. But I found the cause of the failure. If I add a DateTimeValue = new DateTime(2025, 6, 30, 9, 5, 12) it seems to work fine.
Cause is that if I do not set DateTimeValue it defaults to 1/01/0001 0:00:00 with kind = unspecified. The kind = unspecified is interpreted as UTC time so sent to server as 0001-01-01T00:00:00.000Z. The AddAsync call then returns this correctly as 0001-01-01T00:00:00.000Z back to the client, but there the DateTimeConverter class does a DateTime.Parse() which converts the DateTime to local time kind and adjust the time to the local time zone.
The BeEquivalentTo call then compares the original datetime in unspecified timezone with the new one in local time zone and fails because it does not seem to look at the datetime kind.
Not sure if this needs fixing and how to fix it.

powerdude commented on src/CommunityToolkit.Datasync.Client/Offline/SynchronizationEventArgs.cs:
```diff
@@ -0,0 +1,97 @@
+    /// On pull events, reporting occurs per entity type. With a start/stop per entity type.
+    /// On push events, reporting occurs per push request, which may contain multiple entity types.
+    /// </remarks>
+    public required SynchronizationEventType EventType { get; init; }
```
Could this be called Type and the other Type be called EntityType?

StefanCuypers commented:
@adrianhall : the test seems to fail on a timeout on QueueHandler_WithThreads_EnqueueRange(8). If I run this test on my PC it runs in 6sec which is well within the 10s limit. Maybe this was a temporary issue on the test servers? Could you rerun the test?

adrianhall commented:
There is an issue about "flaky tests" - this is one of those cases where the underlying system can decide to go slow and we are left at its mercy.  The only way to resolve it is to perhaps retry the test a few times, or extend the timeout.  

There is also a common problem where the database container we are using has clock drift from the main test suite.  This is likely not the case here though since I believe it's all in the client, so we aren't worried about the timestamps on the server side.

adrianhall commented on src/CommunityToolkit.Datasync.Client/Offline/SynchronizationEventArgs.cs:
```diff
@@ -0,0 +1,97 @@
+    /// When pulling records, the total number of items to pull for the given entity in the current pull request
+    /// When pushing records, the total number of items that are being pushed in the current push request.
+    /// </summary>
+    public long TotalNrItems { get; init; }
```
I agree - TotalItemCount maybe?
